File: python/oelMVCS/Core/Any.py
# -*- coding: utf-8 -*-

from enum import Enum

class Any:
    """
    基础值的任意类型
    """

    class Tag(Enum):
        NULL = 0
        StringValue = 1
        IntValue = 2
        LongValue = 3
        FloatValue = 4
        DoubleValue = 5
        BoolValue = 6

    def __init__(self):
        self.__value = None
        self.__tag:Tag = Tag.Null

    def FromString(_value:str)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.StringValue
        return newAny

    def FromFloat(_value:float)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.FloatValue
        return newAny

    def FromDouble(_value:float)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.DoubleValue
        return newAny

    def FromBool(_value:bool)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.BoolValue
        return newAny

    def FromInt(_value:int)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.IntValue
        return newAny

    def FromLong(_value:long)
        newAny = Any()
        newAny.__value = _value
        newAny.__Tag = Tag.LongValue
        return newAny

    def IsNull(self) -> bool:
        return self.__tag == Tag.NULL

    def IsString(self) -> bool:
        return self.__tag == Tag.StringValue

    def IsInt(self) -> bool:
        return self.__tag == Tag.IntValue

    def IsLong(self) -> bool:
        return self.__tag == Tag.LongValue

    def IsFloat(self) -> bool:
        return self.__tag == Tag.FloatValue

    def IsDouble(self) -> bool:
        return self.__tag == Tag.DoubleValue

    def IsBool(self) -> bool:
        return self.__tag == Tag.BoolValue

    def AsString(self) -> str:
        return self.__value

    def AsInt(self) -> int:
        return self.__value

    def AsLong(self) -> long:
        return self.__value

    def AsFloat(self) -> float:
        return self.__value

    def AsDouble(self) -> double:
        return self.__value

    def AsBool(self) -> bool:
        return self.__value

File: unity/oelMVCS/Assets/XTC/oelMVCS/Scripts/Core/Any.cs

namespace XTC.oelMVCS
{
    public class Any
    {
        public enum Tag
        {
            NULL = 0,
            StringValue = 1,
            IntValue = 2,
            LongValue = 3,
            FloatValue = 4,
            DoubleValue = 5,
            BoolValue = 6
        }

        private string value_ = "";
        private Tag tag_ = Tag.NULL;

        public Any()
        {
        }

        public static Any FromString(string _value)
        {
            Any any = new Any();
            any.tag_ = Tag.StringValue;
            any.value_ = _value;
            return any;
        }

        public static Any FromFloat(float _value)
        {
            Any any = new Any();
            any.tag_ = Tag.FloatValue;
            any.value_ = _value.ToString();
            return any;
        }

        public static Any FromDouble(double _value)
        {
            Any any = new Any();
            any.tag_ = Tag.DoubleValue;
            any.value_ = _value.ToString();
            return any;
        }

        public static Any FromBool(bool _value)
        {
            Any any = new Any();
            any.tag_ = Tag.BoolValue;
            any.value_ = _value.ToString();
            return any;
        }

        public static Any FromInt(int _value)
        {
            Any any = new Any();
            any.tag_ = Tag.IntValue;
            any.value_ = _value.ToString();
            return any;
        }

        public static Any FromLong(long _value)
        {
            Any any = new Any();
            any.tag_ = Tag.LongValue;
            any.value_ = _value.ToString();
            return any;
        }

        public bool IsNull()
        {
            return tag_ == Tag.NULL;
        }

        public bool IsString()
        {
            return tag_ == Tag.StringValue;
        }

        public bool IsInt()
        {
            return tag_ == Tag.IntValue;
        }

        public bool IsLong()
        {
            return tag_ == Tag.LongValue;
        }

        public bool IsFloat()
        {
            return tag_ == Tag.FloatValue;
        }

        public bool IsDouble()
        {
            return tag_ == Tag.DoubleValue;
        }

        public bool IsBool()
        {
            return tag_ == Tag.BoolValue;
        }

        public string AsString()
        {
            return value_;
        }


        public int AsInt()
        {
            int v = 0;
            if (int.TryParse(value_, out v))
                return v;
            return 0;
        }

        public long AsLong()
        {
            long v = 0;
            if (long.TryParse(value_, out v))
                return v;
            return 0;
        }

        public float AsFloat()
        {
            float v = 0.0f;
            if (float.TryParse(value_, out v))
                return v;
            return 0.0f;
        }

        public double AsDouble()
        {
            double v = 0.0;
            if (double.TryParse(value_, out v))
                return v;
            return 0.0;
        }

        public bool AsBool()
        {
            bool v = false;
            if (bool.TryParse(value_, out v))
                return v;
            return !string.IsNullOrEmpty(value_);
        }
    }//class
}//namespace
